managed/CounterStrikeSharp.API/Modules/Menu/ChatMenu.cs

```diff
@@ -1,164 +1,312 @@
-﻿using System;
+using CounterStrikeSharp.API.Modules.Utils;
 using System.Collections.Generic;
-using System.Runtime.InteropServices;
 using System.Text;
-using CounterStrikeSharp.API;
-using CounterStrikeSharp.API.Core;
-using CounterStrikeSharp.API.Modules.Entities;
-using CounterStrikeSharp.API.Modules.Utils;
-
-namespace CounterStrikeSharp.API.Modules.Menu;
+using static CounterStrikeSharp.API.Core.Listeners;
 
-public class ChatMenuOption
+namespace CounterStrikeSharp.API.Modules.Menu
 {
-    public ChatMenuOption(string text, bool disabled, Action<CCSPlayerController, ChatMenuOption> onSelect)
+    public enum DisplayMethod
     {
-        Text = text;
-        Disabled = disabled;
-        OnSelect = onSelect;
+        Chat,
+        CenterHtml,
+        Center,
+        Console
     }
 
-    public Action<CCSPlayerController, ChatMenuOption> OnSelect { get; set; }
-
-    public string Text { get; set; }
-    public bool Disabled { get; set; }
-}
+    public class ChatMenuOption
+    {
+        public Action<CCSPlayerController, ChatMenuOption> OnSelect { get; set; }
+        public string Text { get; set; }
+        public bool Disabled { get; set; }
 
-public class ChatMenu
-{
-    public string Title { get; set; }
-    public List<ChatMenuOption> MenuOptions { get; } = new();
+        public ChatMenuOption(string text, bool disabled, Action<CCSPlayerController, ChatMenuOption> onSelect)
+        {
+            Text = text;
+            Disabled = disabled;
+            OnSelect = onSelect;
+        }
+    }
 
-    public ChatMenu(string title)
+    public class ChatMenu
     {
-        Title = title;
+        public string Title { get; set; }
+        public List<ChatMenuOption> MenuOptions { get; } = new();
+        public DisplayMethod DisplayMethod { get; private set; }
+
+        public ChatMenu(string title, DisplayMethod displayMethod)
+        {
+            Title = title;
+            DisplayMethod = displayMethod;
+        }
+
+        public ChatMenuOption AddMenuOption(string display, Action<CCSPlayerController, ChatMenuOption> onSelect, bool disabled = false)
+        {
+            var option = new ChatMenuOption(display, disabled, onSelect);
+            MenuOptions.Add(option);
+            return option;
+        }
     }
 
-    public ChatMenuOption AddMenuOption(string display, Action<CCSPlayerController, ChatMenuOption> onSelect, bool disabled = false)
+    public class ChatMenuInstance : BasePlugin

```

Comment:
Is this supposed to be a BasePlugin now?

Reply:
It would be better to pass a `BasePlugin`instance to the method `OpenMenu` in ChatMenus :

```csharp
public static class ChatMenus
{
    public static void OpenMenu(BasePlugin plugin, CCSPlayerController player, ChatMenu menu)
    {
        ActiveMenus[player.Handle] = new ChatMenuInstance(plugin, player, menu);
        ActiveMenus[player.Handle].Display();
    }
}
```

This way, you can use the BasePlugin to register your listener in the constructor and avoid the menu being a plugin itself.

Reply:
Thank you very much for the improvement, updated the code

Reply:
You still need to update ChatMenus static class. Search in your IDE for the class.